designs/2023-only-run-reporting-rules/README.md

```diff
@@ -48,46 +48,62 @@ should verify that the `rulePredicate` option is a function, and replace it with
 function if not. `processOptions` in `eslint-helpers.js` should also perform a validation check
 that the `rulePredicate` option is a function.
 
+The new `rulePredicate` function when implemented would look like this, using the `--quiet` flag
+rules outlined in this RFC as an example:
+
+```typescript
+linter.verifyAndFix(text, configs, {
+    rulePredicate: ({ ruleId: string, rule: Rule, severity: number }) => {
```

Comment:
Maybe `filterRules()` would be clearer than `rulePredicate` here?

Reply:
It looks like this would be called once for each rule -- if someone has hundreds of rules enabled, I'm wondering about a perf hit of this approach vs. having a function that just inspects the entire `rules` section of a config and just returns the pieces that should be run. What do you think?

Reply:
or `ruleFilter`

Reply:
The name change makes sense; will do

My only concern around the performance change there is, it'd still be constructing an array from the configuredRules and grabbing all the data, and then having to re-construct the configuredRules list to later iterate on. There'd be a fair bit of double-ups on data access, creation of an array, and filtering/mapping/etc. I feel that overhead might outweigh function calls here in most cases

Reply:
Sorry, I'm not quite following. What change are you referring to here?

I think running this function over every rule is the biggest perf hit we'll take, so I'd like to see if there's a way to avoid that.

Reply:
I've just pushed a change that switches to using a single filter function, and also updated the patch in the opening post to do this. I haven't personally found any performance differences but I'm sure the actual implementation could be optimised further

Reply:
Just to make sure I'm understanding -- are you saying you did a comparison of running this function once per rule vs. running the function once on the entire rule config?

It looks like the latest version is still a function that requires execution on each rule in the config, and that's the concern I have. I'm very interested in the difference between running `(ruleId, severity, rule) => boolean` over every rule vs running `(rules: { [string]:[severity]}) => boolean` one time. 

@mdjermanovic do you have thoughts around this?

Reply:
On the machines I tested I didn't find any difference, but maybe if there were vastly more rules it'd make more of a difference. 

I'm not quite sure what you mean by the method signature you're wanting here, returning a single boolean value for all rules would just cancel all runs wouldn't it? No matter what is done here, in some way every rule needs to be checked - my current implementation now moves that to an implementation detail so using faster methods like for loops etc is possible.

The implementation in the current document uses one function that takes the entire list and returns a filtered down list (or an identity function), but yes still runs a filter over each rule in that list within the function implementation. While it could be optimised by using a for loop rather than the filter function, I feel given that's an implementation change rather than any impact to the API or functionality that it'd be a micro optimisation that doesn't really make sense to do over clarity in an RFC. I have tried to optimize anywhere in the patch however that doesn't result in worse clarity

Reply:
For a more visual example (https://gist.github.com/me4502/e75e4677d3ff17ad585fcab0ac5dc6e9#file-eslint_rfc_rough_implementation-patch-L213) - the way I've written it currently creates a new list from `configuredRules` that grabs the bits of data that are pulled in early on in the current iteration over that list. This lets it filter out disabled rules, etc beforehand. That new list is passed to the filter function, and the returned list from the filtered function is then iterated over and the rules are run

As it's useful for the filter function to have that data, IMO it makes sense to do it this way

Reply:
Yeah that was my main concern with switching to this style given in any case it'd be checking against every rule and this might be slower due to the extra array allocation (https://github.com/eslint/rfcs/pull/104#discussion_r1113790177), although my original concern of duplicated data access didn't end up being an issue in my implementation. While it could be replaced with a for loop and a new array, I feel I trust the inbuilt filter function here to perform the array side of things faster than writing it in userland would be.

I can't think of a way that this overhead could be entirely negated, aside from allowing this function to be undefined and not calling it when it's not in use. That would remove the overhead in cases where it won't likely lead to a performance improvement, but might complicate the API a little more

Reply:
Sorry, finally getting back to this. What I was thinking is instead of doing this:

```js
{
    ruleFilter(ruleId, rule, severity) {
        return severity === 1;
    }
}
```

In this case, we are running the function once on every rule when two of those parameters are actually already available inside of the rule config (and it seems like we don't have a use case for the second parameter currently?).

We would instead just pass in the rules config and return a filtered one.

```js
{
    ruleFilter(rulesConfig) {
        return Object.fromEntries(Object.entries(rulesConfig).filter(([,options]) => options[0] === 1));
    }
}
```

Reply:
Thanks - I've removed the `rule` option from the RFC document

Reply:
In this case, would you be happy for me to go back to how it was beforehand where it ran per-rule while iterating through the configuredRules array? The current method where it creates a new array from my testing is actually slower